### ISSUE TITLE
WIP: Redesign `cabal {upload,haddock,check,sdist}` to provide a better Cabal/Hackage UX

### ISSUE THREAD

**Issue by hvr** [haskell/cabal#3837]
This ticket is here to remind me to write down a new workflow design involving the commands

 - `cabal sdist`
 - `cabal check`
 - `cabal haddock`
 - `cabal upload`

and their interactions to make it more convenient for users, as well as improve quality of Hackage uploads a bit.

The basic gist is that `cabal upload` shall provide a convenient way to perform the following sequence of commands (and `sdist`/`check`/`haddock` should be augmented consistently):

 - `cabal sdist`
 - unpack the just generated tarball $SRCTAR in a temp folder, and `chdir(2)` to it
 - echo 'packages: .' > cabal.project
 - cabal new-build
 - cabal new-test
 - cabal new-benchmark
 - cabal new-haddock --for-hackage
 - cabal upload $OLDPWD/dist/$SRCTAR
 - cabal upload -d dist/$DOCTAR

This all without sacrificing control over the individual phases if the user needs it.

**Comment by ezyang** [comment]
Sure why not. I don't upload sdists all that often so I can't assess if this fills a need.

**Comment by ezyang** [comment]
@hvr when are you aiming this for? I guess not 2.0? How about 2.2?

**Comment by hvr** [comment]
@ezyang yeah, 2.2 is more realistic; it'd be useful to have more pieces working (new-test/new-haddock) before tackling this one; and also to combine this with tge switch to new-build as default

**Comment by phadej** [comment]
I'm unmilestoning this. The design will arrive when it will arrive. We have bumping this since 2.2.
